### PR TITLE
Total Due/Debt Amount 

### PR DESCRIPTION
**Pull request by WebShells**
Total Due/Debt Amount in Detailed Transactions & Specific Customer Report.

### CODE REVIEW

**Comment by mckaygerhard** [comment]
hi @WebShells this "due" still not has the accounting, so those "totals" are some confusing for users i think.. i can't explain in english but "total deb"? from what? i only have a "due" payment.. 


later that payment can be "changed" as i noted in the reports, .. so the accounting now dissapear..

**Comment by WebShells** [comment]
@mckaygerhard due & debt have exactly the same definition :x ... Accounting is something that needs alot of work in order to be supported... Its not forgotten, but its not the right timing...

**Comment by mckaygerhard** [comment]
umm so i suggest that the type of payment "due" must be read only and cannot be changed! if not there's no sense!

File: application/models/reports/Detailed_sales.php [hunk 1]
```diff
@@ -159,9 +159,14 @@ public function getData(array $inputs)
 
 	public function getSummaryData(array $inputs)
 	{
-		$this->db->select('SUM(subtotal) AS subtotal, SUM(tax) AS tax, SUM(total) AS total, SUM(cost) AS cost, SUM(profit) AS profit');
+		$sql = $this->db->query("SELECT ospos_sales_payments.sale_id, ospos_sales_payments.payment_type, SUM(ospos_sales_payments.payment_amount) AS amount FROM ospos_sales_payments JOIN ospos_sales ON ospos_sales_payments.sale_id = ospos_sales.sale_id WHERE ospos_sales_payments.payment_type = 'Due' AND DATE(ospos_sales.sale_time) BETWEEN '".$inputs['start_date']."' AND '".$inputs['end_date']."'");
```
Comment (daN4cat) [inline application/models/reports/Detailed_sales.php]: If you check the overall models, you will find that there is no mention of ospos_ prefix (I made everything consistent probably a year ago).
The reason being that if you use CI db functions that's added for you and people could change database table prefix if they want.

So I'm afraid you will need to use the functions select, where and etc.

Comment (daN4cat) [reply application/models/reports/Detailed_sales.php]: Also you need to escape the input data when you add it to the SQL query, see the existing where statement: 

> $this->db->escape($inputs['start_date'])

File: application/models/reports/Specific_customer.php [hunk 2]
```diff
@@ -142,15 +142,17 @@ public function getData(array $inputs)
...
+		if ($sql->num_rows() > 0) {
+		// output data of each row
+		foreach ($sql->result_array() as $row)
+		{		 
+		$this->db->select('SUM(subtotal) AS subtotal, SUM(tax) AS tax, SUM(total) AS total, SUM(cost) AS cost, SUM(profit) AS profit, "'.floor($row['amount']).'" AS due');
 		$this->db->from('sales_items_temp');
+		$this->db->where('sale_date BETWEEN '. $this->db->escape($inputs['start_date']). ' AND '. $this->db->escape($inputs['end_date']));
 		$this->db->where('customer_id', $inputs['customer_id']);
 		
-		if($inputs['payment_type'] == 'invoices')
```
Comment (daN4cat) [inline application/models/reports/Specific_customer.php]: This looks changing something else, are you sure this is correct?

@SteveIreland could you please check?

File: application/models/reports/Specific_customer.php [hunk 3]
```diff
@@ -142,15 +142,17 @@ public function getData(array $inputs)
 
 	public function getSummaryData(array $inputs)
 	{
-		$this->db->select('SUM(subtotal) AS subtotal, SUM(tax) AS tax, SUM(total) AS total, SUM(cost) AS cost, SUM(profit) AS profit');
+		$sql = $this->db->query("SELECT ospos_sales_payments.sale_id, ospos_sales_payments.payment_type, SUM(ospos_sales_payments.payment_amount) AS amount FROM ospos_sales_payments JOIN ospos_sales ON ospos_sales_payments.sale_id = ospos_sales.sale_id WHERE ospos_sales_payments.payment_type = 'Due' AND DATE(ospos_sales.sale_time) BETWEEN '".$inputs['start_date']."' AND '".$inputs['end_date']."' AND ospos_sales.customer_id ='" . $inputs['customer_id'] ."'");
```
Comment (daN4cat) [inline application/models/reports/Specific_customer.php]: same comments as above

File: application/models/reports/Detailed_sales.php [hunk 4]
```diff
@@ -159,9 +159,14 @@ public function getData(array $inputs)
...
+		if ($sql->num_rows() > 0) {
+		// output data of each row
+		foreach ($sql->result_array() as $row)
```
Comment (daN4cat) [inline application/models/reports/Detailed_sales.php]: Any idea how many rows you might get?
Looping and querying the database is not the most efficient way.
Maybe it's better to build a temporary table and join the queries to get a final result using just two SQL statements.

File: application/models/reports/Specific_customer.php (hunk 2 above)
Comment (odiea) [reply application/models/reports/Specific_customer.php]: @daN4cat  That was my contribution when I added the filters to specific customer. It does remove the totals amount from Invoices when it is removed.

File: application/models/reports/Detailed_sales.php (hunk 4 above)
Comment (mckaygerhard) [reply application/models/reports/Detailed_sales.php]: well make that for "contributor" without more expert knowledge of CI are totally imposible, dont spect user can made that in easy way.. neither using db forge with CI and joins.. without proper documentation

File: application/models/reports/Specific_customer.php (hunk 3 above)
Comment (mckaygerhard) [reply application/models/reports/Specific_customer.php]: you can adeed a code example how to use prefix without using db forge active record of CI, well make that for "contributor" without more expert knowledge of CI are totally imposible, dont spect user can made that in easy way.. neither using db forge with CI and joins.. without proper documentation

File: application/models/reports/Detailed_sales.php (hunk 1 above)
Comment (mckaygerhard) [reply application/models/reports/Detailed_sales.php]: well @WebShells , for you, developers can adeed a code example how to use prefix without using db forge active record of CI, well make that for "contributor" without more expert knowledge of CI are totally imposible, dont spect user can made that in easy way.. neither using db forge with CI and joins.. without proper documentation

contributions are important but need some "sacrifice time" from developers of course @daN4cat

File: application/language/en-GB/reports_lang.php [hunk 5]
```diff
@@ -121,3 +121,5 @@
 $lang["reports_used"] = "Points Used";
 $lang["reports_work_orders"] = "Work Orders";
 $lang["reports_zero_and_less"] = "Zero and Less";
+$lang["reports_due"] = "Total Due";
+$lang["reports_money"] = "Money Available";
```
[the remark this answers is not in the document]
Comment (mckaygerhard) [reply application/language/en-GB/reports_lang.php]: that point was i rwite two comments ago in this pull.. its confuse in many ways

**Comment by mckaygerhard** [comment]
weell these are my predictions: another pull that never will be merged.. too many changes that require time from developers (due explain to contributors) and time from contributors (to learn that  info that are not yet digerible)

**Comment by odiea** [comment]
@mckaygerhard OSPOS is not a training camp for would be developers. The main programmers are real programmers.

**Comment by mckaygerhard** [comment]
umm @odiea i agree with that, but the programers seems dont have enought time.. so collaboration are need right?

**Comment by daN4cat** [comment]
And what does that mean? We need to avoid reviews and accept any kind of code no matter?
OR
Make all the changes for the people that submit something?

Review are there to give feedback and should be considered a training opportunity.

**Comment by SteveIreland** [comment]
From personal experience the first pull request is always the toughest to get through (for me it was the first three) and can be a bit frustrating.  However, @daN4cat does an excellent job of quality control and we all benefit when the code is clean, crisp and consistently maintained according to a suitable standard.

Since I am an active developer it took a bit of effort to learn to adapt to this standard because my fingers are trained to code to a different standard. I also had to change my IDE once I realized my IDE was changing the source automatically according to its own standards.

So a feel your pain but please hang in there.

**Comment by nerv55** [comment]
Why this is big deal, so the pull request was rejected, ok fix it and send it back, if the code is ok and meet the standart imposed by the principal team **may be will be accepted**, that is the thing in opensource communities, someone have to make reviews about it. and is good opportunity to get free training: how to code, how make commit, how make pull request.

But there is other side too, you make copy of source code, modified what ever you want the way you wanted and use it, no body see it, so no complains. Today with internet anyone can be programmer.

**Comment by SteveIreland** [comment]
Wait! Anyone can be a programmer?  You mean that I spent all that time getting a computer science degree for nothing?  Man, that hurts.  :wink:

**Comment by mckaygerhard** [comment]
@SteveIreland this its not the first pull made by @WebShells ... so none of "first try".. 

that's the point: there's no guidance to make the "try's" only that content i (me) put in the wiki...

and yes, i agree with @nerv55 , today some ignorants call "programmers" to anybody that use guindows.. click click click now a "expert" programer ready to get your bucks!

**Comment by WebShells** [comment]
@mckaygerhard we never push anything to main repo without getting the review of the organization members, that's because there might be things that doesn't fit code standards and collaborators/contributors didn't notice.. Thus we wait for other members approval... Although our rights/permissions as collaborators allow us to merge commits... 
Each of us know what can and cant be done and we respect what everyone did to this project whether a programmer/developer or not... Sometime it takes some additional experience in order to enhance the way you write the code and that's challenging (Y) to me... I guess @SteveIreland agrees?

**Comment by SteveIreland** [comment]
I definitely agree.  A change involves so many parts and I am usually focused on coding for success.  After that is accomplished I am highly motivated to publish.  The necessary steps of coding for failure, insuring the code conforms to standard, (making sure all of the debugging code has been removed), etc. are easy  to overlook .... especially when I've been working on it for hours on end and it's close to midnight and my brain is fried ... sometimes I publish so I can go to sleep after having convinced myself that I "got 'er done".   Then I  do the cleanup the next day after one or more people of pointed out what I didn't get done.

**Comment by mckaygerhard** [comment]
no @WebShells i not taking about direct push... :+1:

File: application/models/reports/Detailed_sales.php (hunk 4 above)
Comment (jekkos) [reply application/models/reports/Detailed_sales.php]: @daN4cat indeed if that executes in n+1 fashion then we might have a performance issue there

**Comment by daN4cat** [comment]
@WebShells do you plan to address any of the review feedback?

**Comment by WebShells** [comment]
@daN4cat for sure will proceed with the changes asap...

**Pull request closed without merging**